### PR TITLE
sql: refactor ScalarType's Eq + fix numeric list scale deficiencies

### PR DESCRIPTION
**Pull request by sploiselle**
@benesch mentioned he'd like `ScalarType`'s `Eq` derived rather than hand-rolled, so this PR does that. While I was mucking around in this code, I realized there were some deficiencies in how we scaled `numeric list`s, which I also fixed.

### CODE REVIEW

**Review by mjibson — COMMENTED** [review]
There's no comment or docs about when to use `base_eq` or `==`, and it looks like you could just type `==` and have something happen? Also, how do you know that you correctly used base_eq instead of `==` which still works, so could have been missed if it needed to be base_eq instead?

Why do we still need the Eq trait?

**Comment by sploiselle** [comment]
@mjibson Ah, I will add more docs to explain these kind of odd semantics. When comparing `ScalarType`s, if you don't care about embedded values (e.g. in `Numeric`), you'd use `base_eq`--but will make sure this is more directly addressed on the function itself.

We need the `Eq` trait because other items that contain `ScalarType` require `Eq`, e.g. `RelationDesc` transitively enforces the requirement. How and where those other structs are compared, I am a little fuzzy on. I think Nikhil knows more.

File: src/repr/src/scalar.rs [hunk 1]
```diff
@@ -934,93 +942,7 @@ impl PartialEq for ScalarType {
...
-            Uuid => state.write_u8(18),
-            Oid => state.write_u8(19),
-            Map {
-                value_type,
-                custom_oid,
-            } => {
-                state.write_u8(20);
-                value_type.hash(state);
-                custom_oid.hash(state);
-            }
```
Comment (benesch) [inline src/repr/src/scalar.rs:1023]: Yessss finally!

**Review by benesch — APPROVED** [review]

**Comment by benesch** [comment]
> We need the `Eq` trait because other items that contain `ScalarType` require `Eq`, e.g. `RelationDesc` transitively enforces the requirement. How and where those other structs are compared, I am a little fuzzy on. I think Nikhil knows more.

Various bits of `transform` and `dataflow` compare expressions for equality or stash them in hash maps to determine whether two expressions are equivalent. This is a little sketchy, but roughly these bits are happy with false negatives (two expressions are actually semantically equivalent but not reported as equal), but wrong if presented with false positives (two expressions are not actually semantically equivalent but reported as equal). So I feel better about going with an overly strict interpretation of equality in the default case, than with one that is possibly too loose.

Thanks very much @sploiselle! I defer to @mjibson on what the right amount of docs are here, because I've largely gotten used to how confusing this all is.

**Comment by sploiselle** [comment]
@mjibson Going to merge this with the additional docs, but if you'd like to see any other changes lmk and I can iterate further.

**Pull request merged**